## Pandas
- Pandas is a open source python library designed specifically for accomplishing data analysis tasks.
- When we look into source code of pandas, we  can verify that Pandas library is built on top of Numpy library.
- Pandas library provides two important data structures named:
    - [pandas.Series:](https://github.com/pandas-dev/pandas/blob/v2.0.3/pandas/core/series.py#L244-L6112)
        - Series is a one-dimensional ndarray with axis labels.
        - You can think of Series as a Single Column or Single Row of data with associated labels.

    - [`pandas.DataFrame:`](https://github.com/pandas-dev/pandas/blob/v2.0.3/pandas/core/frame.py#L490-L11586)
        - A DataFrame is a two-dimensional labeled data structure with columns that can hold data of different types.
        - It is similar to a table in a relational database or a spreadsheet with rows and columns.
        - Each column in a DataFrame can be visualize as Series.
        - A DataFrame can be created from dictionaries, arrays, list of dictionaries, CSV Files, Excel Files, and more.
        - Pandas DataFrame class consists of methods and attributes that makes the life of data analyst easier.

In addition to these Data Structures, Pandas also provides various tools for Data ingestion, Data manipulation, Data Merging, Grouping and so on.


## Install Pandas and Numpy
- Before using pandas, you need to make sure pandas libary is installed into your system.
- If you are using Google Colab, there is no need to install pandas as it comes with by default.
- In addition to Pandas, we will also use Numpy library, lets install this as well (skip if you are using Google Colab)
- Uncomment below cell to install necessary libraries

In [3]:
# !pip install pandas
# !pip install numpy

## Import Libraries
- In order to use advantange of installed libraries i.e. numpy and pandas, We need to import it using `import` keyword.
- We will represent:
    - numpy via alias `np`
    - pandas via alias `pd`

In [4]:
import numpy as np
import pandas as pd

## Creating Series
- There are several ways in creating a Series:
    1. From List
    2. From Dictionary
    3. From Numpy Array

- To create a a Series you can use `pandas.Series()` function.

**1. Create a Series from python list**

In [7]:
# define list
data = [10, 20, 30, 40, 50]

# create series from list
series_list = pd.Series(data)

print(series_list)

0    10
1    20
2    30
3    40
4    50
dtype: int64


In [6]:
# check type
print(type(series_list))

<class 'pandas.core.series.Series'>


**2. Create Series from Dictionary**

In [10]:
# define dictionary
data_dict = {
    'a': 10,
    'b': 20,
    'c': 30,
    'd': 40
}

# create series from dictionary
series_dict = pd.Series(data_dict)

print(series_dict)

a    10
b    20
c    30
d    40
dtype: int64


In [9]:
# check type
print(type(series_dict))

<class 'pandas.core.series.Series'>


**Create Dictionary from Numpy Array**

In [11]:
# define array
numpy_array = np.array([10, 20, 30, 40])

# create series from numpy array
series_np = pd.Series(numpy_array)

print(series_np)

0    10
1    20
2    30
3    40
dtype: int64


- **Note:**
    - When creating `series` via list and  numpy array, Series index label are automatically generated starting from 0.
    - When creating `Series` via dictionary, Dictionary keys becomes Series index label and Dictionary values becomes the data in Series.

## Create DataFrame
- Several ways to create Pandas DataFrame are:
    1. Create DataFrame from 2D List
    2. Create DataFrame a Dictionary of Lists
    3. Create DataFrame from a List of Dictionaries
    4. Create DataFrame from a CSV file
    5. Create DataFrame from an Excel file

- To create a DataFrame you can use `pd.DataFrame()` function.

**1. Create DataFrame from 2D List**

In [12]:
# Define list
data = [[10, 20], [30, 40], [50, 60]]

# create DataFrame
df_list = pd.DataFrame(data, columns=['col1', 'col2'])

df_list

col1  col2
0    10    20
1    30    40
2    50    60

**2. Create DataFrame a Dictionary of Lists**
- You can create a DataFrame from a Dictionary where,
    - `key` represent column names.
    - `value` represent list containing data for each column.

In [13]:
# define dictionary of lists
data_dict = {
    'Name': ['Ram', 'Gita', 'Hari'],
    'Age': [10, 25, 20]
}

# create dataframe
df_dict = pd.DataFrame(data_dict)

df_dict

Name  Age
0   Ram   10
1  Gita   25
2  Hari   20

**3. Create DataFrame from a List of Dictionaries**
- You can also create DataFrame from a List of Dictionaries, where,
    - each dictionary corresponds to a row and contains column-value pairs.

In [14]:
# define data list
data_list = [
    {'Name': 'Ram', 'Age': 10},
    {'Name': 'Gita', 'Age': 25},
    {'Name': 'Hari', 'Age': 20}
]

# Create DataFrame
df_list = pd.DataFrame(data_list)

df_list

Name  Age
0   Ram   10
1  Gita   25
2  Hari   20

## Data Ingestion
- Data Ingestion is the process of obtaining and import data for immediate use or storage in a database.
- This repository includes variety of Dataset, so for now we will use that as data source.
- There may be scenario where we need to ingest data from Google Drive as well for that purpose we can use `gdown` library

**5. Create DataFrame from Csv File**
- Use `pd.read_csv()` function

In [24]:
# define path
data_path = '../../Data/HR_comma_sep.csv'

# read data from csv file
df_hr = pd.read_csv(data_path)

df_hr

satisfaction_level  last_evaluation  number_project  \
0                    0.38             0.53               2   
1                    0.80             0.86               5   
2                    0.11             0.88               7   
3                    0.72             0.87               5   
4                    0.37             0.52               2   
...                   ...              ...             ...   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                       157                   3              0     1   
1                       262                   6              0     1   
2                       272                   4              0     1   
3                       223                   5              0     1   
4                       159                   3              0     1   
...                     ...                 ...            ...   ...   
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department  salary  
0                          0      sales     low  
1                          0      sales  medium  
2                          0      sales  medium  
3                          0      sales     low  
4                          0      sales     low  
...                      ...        ...     ...  
14994                      0    support     low  
14995                      0    support     low  
14996                      0    support     low  
14997                      0    support     low  
14998                      0    support     low  

[14999 rows x 10 columns]

**5. Create DataFrame from an Excel file**
- Use  `pd.read_excel()` function

In [ ]:
# write your code here

## Viewing Data
- In this sections we will explore methods available to view our loaded DataFrame object.
- **Methods are**
    - head() and tail()
    - shape
    - index and columns
    - dtypes
    - info()  
    - describe()

In [26]:
# use head() to view top 5 rows
df_hr.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [27]:
# use tail() to view last 5 rows
df_hr.tail()

satisfaction_level  last_evaluation  number_project  \
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department salary  
14994                      0    support    low  
14995                      0    support    low  
14996                      0    support    low  
14997                      0    support    low  
14998                      0    support    low

In [32]:
# Check dataframe shape
# Hint: df.shape

df_hr.shape

(14999, 10)

In [30]:
# extract dataframe index
# Hint: df.index

df_hr.index

RangeIndex(start=0, stop=14999, step=1)

In [31]:
# extract  columns
# Hint: df.columns

df_hr.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'Department', 'salary'],
      dtype='object')

In [34]:
# get column types
# Hint: df.dtypes

df_hr.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
left                       int64
promotion_last_5years      int64
Department                object
salary                    object
dtype: object

- We can see, 
    - 2 columns are of type float64
    - 7 columns are of type int64
    - 2 columns are of type object

In [35]:
# get basic dataframe information
# Hint: df.info()

df_hr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   Department             14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


- Dataframe occupies 1.1+ MB memory
- Columns are of types: float64, int64, object
- Total rows = 14999
- No any columns contains null values

In [37]:
# Descriptive Statistics
# Hint: df.describe()

df_hr.describe()

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

- We get the summary of each dataframe column easily without the need for explicit computation.
- By default it includes only the numerical columns.
- let's get description  of object type only in below cells.

In [52]:
# describe object type, exclude int, and float
# Hint: df.describe(exclude=['int', 'float'])

df_hr.describe(exclude=['int', 'float'])

Department salary
count       14999  14999
unique         10      3
top         sales    low
freq         4140   7316

- For Object type, we get count, unique, top items with frequency.

## Data Selection
- This section covers different techniques to access columns, slice dataframe, get unique column values.
    - select columns (using [], and dot(.) operator)
    - iloc
    - loc 
    - unique() and nunique()
    - value_counts()

**Select Columns**

In [42]:
# Select single column

# without using dot operator
print(df_hr['satisfaction_level'])

print('------------------------------------------')

# using dot operator
print(df_hr.satisfaction_level)

0        0.38
1        0.80
2        0.11
3        0.72
4        0.37
         ... 
14994    0.40
14995    0.37
14996    0.37
14997    0.11
14998    0.37
Name: satisfaction_level, Length: 14999, dtype: float64
------------------------------------------
0        0.38
1        0.80
2        0.11
3        0.72
4        0.37
         ... 
14994    0.40
14995    0.37
14996    0.37
14997    0.11
14998    0.37
Name: satisfaction_level, Length: 14999, dtype: float64


In [45]:
# Select multiple columns

columns_to_select = ['satisfaction_level', 'left']

df_hr[columns_to_select]

satisfaction_level  left
0                    0.38     1
1                    0.80     1
2                    0.11     1
3                    0.72     1
4                    0.37     1
...                   ...   ...
14994                0.40     1
14995                0.37     1
14996                0.37     1
14997                0.11     1
14998                0.37     1

[14999 rows x 2 columns]

**iloc**  
- ways to slice pandas dataframe via integer value.
- It allows you to select data from a DataFrame using integer indices for both rows and columns.
- 1st row is 0, 2nd row is 1, and so on.
- 1st column is 0, 2nd column is 1, and so on.
- **`syntax: df.iloc[row_index, column_index]`**

In [47]:
# select single element using iloc
# value at 1st row and 1st column

df_hr.iloc[0, 0]

0.38

In [48]:
# slice dataframe
# 1st 2 rows and last 3 columns

df_hr.iloc[:3, -3:]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium
2                      0      sales  medium

**loc**
- `loc` is used for label-based indexing. 
- Allows us to slice dataframe using row label and column label instead of integer value as in loc.  
- **`syntax: df.loc[row_label, column_label]`**

In [49]:
# select single element using loc
# value at 1st row and 1st column

df_hr.loc[0, 'satisfaction_level']

0.38

In [51]:
# slice dataframe
# 1st 2 rows and last 3 columns

row_label = [0, 1, 2]
column_label = ['promotion_last_5years', 'Department', 'salary']

df_hr.loc[row_label, column_label]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium
2                      0      sales  medium

**unique() and nunique()**
- `unique()` function to get unique item names. You need to have Series data structure not DataFrame.
- `nunique()` function to get number of unique item names. You need to have Series data structure not DataFrame.

In [54]:
# get unique item names of column "left"
# Hint: df['col_name'].unique()

df_hr['left'].unique()

array([1, 0])

In [55]:
# get unique item count of column "left"
# Hint: df['col_name'].nunique()

df_hr['left'].nunique()

2

- This means `left` column has 2 unique values,
    - 1: indicates employee leaving company
    - 0: indicates employee staying in company

## Aggregations
- `Methods are:`
    - mean()
    - std()
    - sum()
    - max()
    - min()

## Missing Data
- `Methods are:`
    - isnull() or isna()
    - notnull() or notna()
    - dropna()
    - fillna()

## Data Manipulation
- `Methods are:`
    - replace()
    - astype()
    - drop()
    - sort_values()
    - groupby()
    - apply()
    

## Data Join
- `Types are:`
    1. Inner Join
        - pd.merge(df1, df2, on='col_name', how='inner')
    2. Left Outer Join
        - pd.merge(df1, df2, on='col_name', how='left')
    3. Right Outer Join
        - pd.merge(df1, df2, on='col_name', how='right')
    4. Full Outer Join
        - pd.merge(df1, df2, on='col_name', how='outer')

## Save Data
- Like Read, Pandas provides several function to save data to various file formats such as CSV, Excel, JSON, and so on.
- **Save CSV**
    - `df.to_csv(csv_file_path, index=False)`
        - csv_file_path: csv path where file needs to be saved: e.g. test.csv
        - index=False: means do not save index data.

- **Save Excel**
    - `df.to_excel(excel_file_path, index=False)`

In [25]:
# write your program here